# Pandas Project 1


E-Commerce Analysis Using Pandas

In [3]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("zusmani/pakistans-largest-ecommerce-dataset")

print("Path to dataset files:", path)

Using Colab cache for faster access to the 'pakistans-largest-ecommerce-dataset' dataset.
Path to dataset files: /kaggle/input/pakistans-largest-ecommerce-dataset


In [4]:
# Getting the data
import pandas as pd
data = pd.read_csv('/kaggle/input/pakistans-largest-ecommerce-dataset/Pakistan Largest Ecommerce Dataset.csv')

/tmp/ipython-input-2268948685.py:3: DtypeWarning: Columns (1,2,3,7,8,9,11,12,13,14,17,18,19) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv('/kaggle/input/pakistans-largest-ecommerce-dataset/Pakistan Largest Ecommerce Dataset.csv')


In [5]:
# Converting the data to a pandas dataframe
df = pd.DataFrame(data)
df.head()

,item_id,status,created_at,sku,price,qty_ordered,grand_total,increment_id,category_name_1,sales_commission_code,...,Month,Customer Since,M-Y,FY,Customer ID,Unnamed: 21,Unnamed: 22,Unnamed: 23,Unnamed: 24,Unnamed: 25
0,211131.0,complete,7/1/2016,kreations_YI 06-L,1950.0,1.0,1950.0,100147443,Women's Fashion,\N,...,7.0,2016-7,7-2016,FY17,1.0,NaN,NaN,NaN,NaN,NaN
1,211133.0,canceled,7/1/2016,kcc_Buy 2 Frey Air Freshener & Get 1 Kasual Bo...,240.0,1.0,240.0,100147444,Beauty & Grooming,\N,...,7.0,2016-7,7-2016,FY17,2.0,NaN,NaN,NaN,NaN,NaN
2,211134.0,canceled,7/1/2016,Ego_UP0017-999-MR0,2450.0,1.0,2450.0,100147445,Women's Fashion,\N,...,7.0,2016-7,7-2016,FY17,3.0,NaN,NaN,NaN,NaN,NaN
3,211135.0,complete,7/1/2016,kcc_krone deal,360.0,1.0,60.0,100147446,Beauty & Grooming,R-FSD-52352,...,7.0,2016-7,7-2016,FY17,4.0,NaN,NaN,NaN,NaN,NaN
4,211136.0,order_refunded,7/1/2016,BK7010400AG,555.0,2.0,1110.0,100147447,Soghaat,\N,...,7.0,2016-7,7-2016,FY17,5.0,NaN,NaN,NaN,NaN,NaN


In [6]:
# Checking the null values
df.isnull().sum()

,0
item_id,464051
status,464066
created_at,464051
sku,464071
price,464051
qty_ordered,464051
grand_total,464051
increment_id,464051
category_name_1,464215
sales_commission_code,601229


In [7]:
# Exploring the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 26 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   item_id                584524 non-null  float64
 1   status                 584509 non-null  object 
 2   created_at             584524 non-null  object 
 3   sku                    584504 non-null  object 
 4   price                  584524 non-null  float64
 5   qty_ordered            584524 non-null  float64
 6   grand_total            584524 non-null  float64
 7   increment_id           584524 non-null  object 
 8   category_name_1        584360 non-null  object 
 9   sales_commission_code  447346 non-null  object 
 10  discount_amount        584524 non-null  float64
 11  payment_method         584524 non-null  object 
 12  Working Date           584524 non-null  object 
 13  BI Status              584524 non-null  object 
 14   MV                    584524 non-

In [8]:
df.describe()

,item_id,price,qty_ordered,grand_total,discount_amount,Year,Month,Customer ID,Unnamed: 21,Unnamed: 22,Unnamed: 23,Unnamed: 24,Unnamed: 25
count,584524.000000,5.845240e+05,584524.000000,5.845240e+05,584524.000000,584524.000000,584524.000000,584513.000000,0.0,0.0,0.0,0.0,0.0
mean,565667.074218,6.348748e+03,1.296388,8.530619e+03,499.492775,2017.044115,7.167654,45790.511965,NaN,NaN,NaN,NaN,NaN
std,200121.173648,1.494927e+04,3.996061,6.132081e+04,1506.943046,0.707355,3.486305,34414.962389,NaN,NaN,NaN,NaN,NaN
min,211131.000000,0.000000e+00,1.000000,-1.594000e+03,-599.500000,2016.000000,1.000000,1.000000,NaN,NaN,NaN,NaN,NaN
25%,395000.750000,3.600000e+02,1.000000,9.450000e+02,0.000000,2017.000000,4.000000,13516.000000,NaN,NaN,NaN,NaN,NaN
50%,568424.500000,8.990000e+02,1.000000,1.960400e+03,0.000000,2017.000000,7.000000,42856.000000,NaN,NaN,NaN,NaN,NaN
75%,739106.250000,4.070000e+03,1.000000,6.999000e+03,160.500000,2018.000000,11.000000,73536.000000,NaN,NaN,NaN,NaN,NaN
max,905208.000000,1.012626e+06,1000.000000,1.788800e+07,90300.000000,2018.000000,12.000000,115326.000000,NaN,NaN,NaN,NaN,NaN


In [9]:
# Dropping the unecessary columns
df = df.drop(["sales_commission_code", "FY", "Unnamed: 21", "Unnamed: 22", "Unnamed: 23", "Unnamed: 24", "Unnamed: 25"], axis=1)
df

,item_id,status,created_at,sku,price,qty_ordered,grand_total,increment_id,category_name_1,discount_amount,payment_method,Working Date,BI Status,MV,Year,Month,Customer Since,M-Y,Customer ID
0,211131.0,complete,7/1/2016,kreations_YI 06-L,1950.0,1.0,1950.0,100147443,Women's Fashion,0.0,cod,7/1/2016,#REF!,"1,950",2016.0,7.0,2016-7,7-2016,1.0
1,211133.0,canceled,7/1/2016,kcc_Buy 2 Frey Air Freshener & Get 1 Kasual Bo...,240.0,1.0,240.0,100147444,Beauty & Grooming,0.0,cod,7/1/2016,Gross,240,2016.0,7.0,2016-7,7-2016,2.0
2,211134.0,canceled,7/1/2016,Ego_UP0017-999-MR0,2450.0,1.0,2450.0,100147445,Women's Fashion,0.0,cod,7/1/2016,Gross,"2,450",2016.0,7.0,2016-7,7-2016,3.0
3,211135.0,complete,7/1/2016,kcc_krone deal,360.0,1.0,60.0,100147446,Beauty & Grooming,300.0,cod,7/1/2016,Net,360,2016.0,7.0,2016-7,7-2016,4.0
4,211136.0,order_refunded,7/1/2016,BK7010400AG,555.0,2.0,1110.0,100147447,Soghaat,0.0,cod,7/1/2016,Valid,"1,110",2016.0,7.0,2016-7,7-2016,5.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1048570,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1048571,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1048572,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1048573,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [10]:
# The dataset contains Item ID, Order Status (Completed, Cancelled, Refund), Date of Order, Stock Keeping Unit, Price, Quantity, Grand Total, Category, Payment Method and Customer ID.
df.head()

,item_id,status,created_at,sku,price,qty_ordered,grand_total,increment_id,category_name_1,discount_amount,payment_method,Working Date,BI Status,MV,Year,Month,Customer Since,M-Y,Customer ID
0,211131.0,complete,7/1/2016,kreations_YI 06-L,1950.0,1.0,1950.0,100147443,Women's Fashion,0.0,cod,7/1/2016,#REF!,"1,950",2016.0,7.0,2016-7,7-2016,1.0
1,211133.0,canceled,7/1/2016,kcc_Buy 2 Frey Air Freshener & Get 1 Kasual Bo...,240.0,1.0,240.0,100147444,Beauty & Grooming,0.0,cod,7/1/2016,Gross,240,2016.0,7.0,2016-7,7-2016,2.0
2,211134.0,canceled,7/1/2016,Ego_UP0017-999-MR0,2450.0,1.0,2450.0,100147445,Women's Fashion,0.0,cod,7/1/2016,Gross,"2,450",2016.0,7.0,2016-7,7-2016,3.0
3,211135.0,complete,7/1/2016,kcc_krone deal,360.0,1.0,60.0,100147446,Beauty & Grooming,300.0,cod,7/1/2016,Net,360,2016.0,7.0,2016-7,7-2016,4.0
4,211136.0,order_refunded,7/1/2016,BK7010400AG,555.0,2.0,1110.0,100147447,Soghaat,0.0,cod,7/1/2016,Valid,"1,110",2016.0,7.0,2016-7,7-2016,5.0


In [11]:
# Checking the null values now
df.isnull().sum()

,0
item_id,464051
status,464066
created_at,464051
sku,464071
price,464051
qty_ordered,464051
grand_total,464051
increment_id,464051
category_name_1,464215
discount_amount,464051


In [12]:
# Dropping the null values because we have a very large dataset
df.dropna(inplace=True)

In [13]:
# Again confirming the dataframe doesn't contains any null values
df.isnull().sum()

,0
item_id,0
status,0
created_at,0
sku,0
price,0
qty_ordered,0
grand_total,0
increment_id,0
category_name_1,0
discount_amount,0


In [14]:
df.shape

(584314, 19)

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 584314 entries, 0 to 584523
Data columns (total 19 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   item_id          584314 non-null  float64
 1   status           584314 non-null  object 
 2   created_at       584314 non-null  object 
 3   sku              584314 non-null  object 
 4   price            584314 non-null  float64
 5   qty_ordered      584314 non-null  float64
 6   grand_total      584314 non-null  float64
 7   increment_id     584314 non-null  object 
 8   category_name_1  584314 non-null  object 
 9   discount_amount  584314 non-null  float64
 10  payment_method   584314 non-null  object 
 11  Working Date     584314 non-null  object 
 12  BI Status        584314 non-null  object 
 13   MV              584314 non-null  object 
 14  Year             584314 non-null  float64
 15  Month            584314 non-null  float64
 16  Customer Since   584314 non-null  object 
 

## Sales & Orders Insights

Total Number of Orders

In [16]:
total_orders = len(df)
print("Total Number of Orders:", total_orders)

Total Number of Orders: 584314


Total Revenue

In [17]:
total_revenue = df["grand_total"].sum()
print("Total Revenue:", total_revenue.round(1))

Total Revenue: 4985888471.3


Average Order Value

In [18]:
avg_order_value = total_revenue / total_orders
print("Average Order Value:", avg_order_value.round(1))

Average Order Value: 8532.9


Orders by Year

In [19]:
yearly_orders = df.groupby(df["Year"]).size().sort_values(ascending=False)
yearly_orders

,0
Year,
2017.0,290908
2018.0,159505
2016.0,133901


Orders by Months

In [20]:
monthly_orders = df.groupby(df["Month"]).size().sort_values(ascending=False)
monthly_orders

,0
Month,
11.0,155453
5.0,62571
3.0,61470
8.0,48512
7.0,39146
2.0,38763
6.0,34526
4.0,33985
10.0,30619


Yearly Revenue

In [21]:
yearly_revenue = df.groupby(df["Year"])["grand_total"].sum().sort_values(ascending=False)
yearly_revenue

,grand_total
Year,
2017.0,2.247206e+09
2018.0,2.093097e+09
2016.0,6.455853e+08


Monthly Revenue

In [22]:
yearly_revenue = df.groupby(df["Month"])["grand_total"].sum().sort_values(ascending=False)
yearly_revenue

,grand_total
Month,
11.0,1.067833e+09
5.0,7.071842e+08
3.0,5.602699e+08
6.0,4.725497e+08
2.0,4.082150e+08
7.0,3.894224e+08
8.0,3.473219e+08
4.0,2.728163e+08
10.0,2.175180e+08


## Product / Category Analysis

Top Product Categories by Orders

In [23]:
cat_by_orders = df.groupby("category_name_1").size().sort_values(ascending=False)
cat_by_orders

,0
category_name_1,
Mobiles & Tablets,115709
Men's Fashion,92218
Women's Fashion,59720
Appliances,52413
Superstore,43611
Beauty & Grooming,41493
Soghaat,34011
Others,29208
Home & Living,26504


Top Product Categories by Revenue

In [24]:
rev_by_cat = df.groupby("category_name_1")["grand_total"].sum().sort_values(ascending=False)
rev_by_cat

,grand_total
category_name_1,
Mobiles & Tablets,2.440777e+09
Appliances,6.568497e+08
Entertainment,5.389914e+08
Women's Fashion,2.825768e+08
Computing,2.025457e+08
Men's Fashion,1.941367e+08
Others,1.924014e+08
Superstore,1.121146e+08
Beauty & Grooming,9.718264e+07


## Payment Method Analysis

Most popular payment methods

In [25]:
payment_method = df.groupby("payment_method").size().sort_values(ascending=False)
payment_method

,0
payment_method,
cod,271770
Payaxis,97640
Easypay,82893
jazzwallet,35145
easypay_voucher,31175
bankalfalah,23057
jazzvoucher,15633
Easypay_MA,14026
customercredit,7554


Revenue by payment method

In [29]:
rev_by_payment_method = df.groupby("payment_method")["grand_total"].sum().sort_values(ascending=False)
rev_by_payment_method.to_frame().style.format('{0:,.0f}')

,grand_total
payment_method,
Payaxis,"1,139,557,936"
cod,"1,065,026,207"
Easypay,"916,213,965"
bankalfalah,"688,263,657"
easypay_voucher,"560,887,810"
jazzvoucher,"251,880,698"
jazzwallet,"162,608,373"
Easypay_MA,"108,904,968"
apg,"27,066,733"


Refund rate by payment method

In [38]:
refunded_order_count = df[df["status"]=='order_refunded'].groupby(df["payment_method"]).size()
refunded_order_rate = (refunded_order_count / refunded_order_count.sum()) * 100
refunded_order_rate.sort_values(ascending=False)

,0
payment_method,
cod,81.925443
Payaxis,6.408619
Easypay,4.405190
customercredit,2.131164
jazzwallet,1.403409
bankalfalah,1.252143
easypay_voucher,0.806750
jazzvoucher,0.692460
Easypay_MA,0.527749


Completion rate by payment method

In [45]:
comp_order_count = df[df["status"]=='complete'].groupby(df["payment_method"]).size()
comp_order_rate = (comp_order_count / comp_order_count.sum()) * 100
comp_order_rate.sort_values(ascending=False)

,0
payment_method,
cod,63.349652
Payaxis,9.760617
Easypay,8.222215
easypay_voucher,6.875096
jazzwallet,5.779172
jazzvoucher,1.976601
customercredit,1.776330
Easypay_MA,1.333425
cashatdoorstep,0.288424


## Customer Behavior

Order frequency by customer

In [42]:
order_freq = df.groupby("Customer ID").size().sort_values(ascending=False)
order_freq

,0
Customer ID,
85775.0,2524
163.0,2349
35.0,1877
33.0,1397
31025.0,1369
...,...
115287.0,1
115286.0,1
115285.0,1


## E-Commerce Data Analysis Insights

### Sales & Orders Insights

*   **Total Number of Orders**: 584,314
*   **Total Revenue**: PKR 4,985,888,471.3
*   **Average Order Value**: PKR 8,532.9

**Orders by Year**:
*   **2017**: 290,908 orders
*   **2018**: 159,505 orders
*   **2016**: 133,901 orders

**Orders by Month** (Top 3):
*   **November**: 155,453 orders
*   **May**: 62,571 orders
*   **March**: 61,470 orders

**Yearly Revenue**:
*   **2017**: PKR 2,247,206,000
*   **2018**: PKR 2,093,097,000
*   **2016**: PKR 645,585,300

**Monthly Revenue** (Top 3):
*   **November**: PKR 1,067,833,000
*   **May**: PKR 707,184,200
*   **March**: PKR 560,269,900

### Product / Category Analysis

**Top Product Categories by Orders** (Top 5):
*   **Mobiles & Tablets**: 115,709 orders
*   **Men's Fashion**: 92,218 orders
*   **Women's Fashion**: 59,720 orders
*   **Appliances**: 52,413 orders
*   **Superstore**: 43,611 orders

**Top Product Categories by Revenue** (Top 5):
*   **Mobiles & Tablets**: PKR 2,440,777,000
*   **Appliances**: PKR 656,849,700
*   **Entertainment**: PKR 538,991,400
*   **Women's Fashion**: PKR 282,576,800
*   **Computing**: PKR 202,545,700

### Payment Method Analysis

**Most Popular Payment Methods** (Top 3 by orders):
*   **COD (Cash on Delivery)**: 271,770 orders
*   **Payaxis**: 97,640 orders
*   **Easypay**: 82,893 orders

**Revenue by Payment Method** (Top 3):
*   **Payaxis**: PKR 1,139,558,000
*   **COD**: PKR 1,065,026,000
*   **Easypay**: PKR 916,214,000

**Refund Rate by Payment Method** (Top 3):
*   **COD**: 17.94% (48744 refunds / 271770 total COD orders)
*   **Customer Credit**: 16.79% (1268 refunds / 7554 total Customer Credit orders)
*   **Bankalfalah**: 3.23% (745 refunds / 23057 total Bankalfalah orders)

**Completion Rate by Payment Method** (Top 3):
*   **COD**: 54.43% (148038 completed / 271770 total COD orders)
*   **Payaxis**: 23.36% (22809 completed / 97640 total Payaxis orders)
*   **Easypay**: 23.18% (19214 completed / 82893 total Easypay orders)

### Customer Behavior

**Top Customers by Order Frequency**:
*   Customer ID `85775.0`: 2,524 orders
*   Customer ID `163.0`: 2,349 orders
*   Customer ID `35.0`: 1,877 orders

(Note: This analysis covers the available data after cleaning and dropping null values.)